shared/modules/Storage/ExtensionStore.ts

```diff
@@ -0,0 +1,210 @@
+import browser from 'webextension-polyfill';
+import log from 'loglevel';
+import { captureException } from '@sentry/browser';
+import { checkForLastError } from '../browser-runtime.utils';
+import type Migrator from '../../../app/scripts/lib/migrator';
+import {
+  type EmptyState,
+  type IntermediaryStateType,
+  type MetaMaskStorageStructure,
+  BaseStorage,
+} from './Storage';
+
+/**
+ * Returns whether or not the given object contains no keys
+ *
+ * @param obj - The object to check
+ * @returns
+ */
+function isEmpty(
+  obj: MetaMaskStorageStructure | EmptyState,
+): obj is EmptyState {
+  return Object.keys(obj).length === 0;
+}
+
+/**
+ * An implementation of the MetaMask Extension Storage system that uses the
+ * browser.storage.local API to persist and retrieve state.
+ */
+export class ExtensionStore extends BaseStorage {
+  isSupported: boolean;
+
+  stateCorruptionDetected: boolean;
+
+  dataPersistenceFailing: boolean;
+
+  mostRecentRetrievedState: MetaMaskStorageStructure | null;
+
+  migrator: Migrator;
+
+  constructor({ migrator }: { migrator: Migrator }) {
+    super();
+    this.stateCorruptionDetected = false;
+    this.isSupported = Boolean(browser.storage.local);
+    if (!this.isSupported) {
+      log.error('Storage local API not available.');
+    }
+    // we use this flag to avoid flooding sentry with a ton of errors:
+    // once data persistence fails once and it flips true we don't send further
+    // data persistence errors to sentry
+    this.dataPersistenceFailing = false;
+    this.mostRecentRetrievedState = null;
+    this.migrator = migrator;
+  }
+
+  /**
+   * Persists the MetaMask state tree on the 'data' key of the
+   * browser.storage.local API. This function will first do some sanity checks
+   * to determine if it is likely for the operation to succeed. It will throw
+   * an error in the following cases.
+   * 1. The browser does not support the browser.storage.local API.
+   * 2. No state object was provided to the function. As more of the codebase
+   * is migrated to TypeScript, this should become less of a possibility but
+   * never impossible.
+   * 3. The metadata property is not set on the class which is required before
+   * setting state. This ensures the 'meta' key of the storage is set.
+   *
+   * @param state - The state to persist to the data key of the local store
+   * @returns void
+   */
+  async set(state: IntermediaryStateType) {
+    if (!this.isSupported) {
+      throw new Error(
+        'Metamask- cannot persist state to local store as this browser does not support this action',
+      );
+    }
+    if (!state) {
+      throw new Error('MetaMask - updated state is missing');
+    }
+    if (!this.metadata) {
+      throw new Error(
+        'MetaMask - metadata must be set on instance of ExtensionStore before calling "set"',
+      );
+    }
+    if (
+      this.stateCorruptionDetected &&
+      window.localStorage.getItem('USER_OPTED_IN_TO_RESTORE') !== 'true'
```

Comment:
question: where is this variable set in storage?

Reply:
This code should be removed from this PR, as its part of the next PR, however to answer your question this is the browsers local storage system which can be viewed in the application -> storage -> local portion of devtools in the background/service worker process. 

Its stored here to overcome the limitations on extension long term storage and the potential of that storage system to become corrupted when read/write processes are interrupted by an irregular browser shutdown process.

Reply:
Ahh what i meant was,  I didn't see where the storage was set, and it is helpful to new it will be removed in next PR 🙏🏻  and it makes sense here to remove long term storage

Reply:
Sorry, no it'll be removed now and added to next PR where it is used. 

This .localStorage is separate and distinct from chrome.storage.local. The latter is a disk file storage for extension data, with the former being the default browser storage. The idea is that .localStorage will not suffer from the same conditions of corruption as chrome.storage.local and thus can be used even if the extension storage system is corrupted.